Ubuntu20 redash

RobHooper opened the pull request:
This PR updates salt to deploy redash on the new ocp14 server running ubuntu 20.04 #145

I have also taken this opportunity to move postgres out of docker and upgrade it #299. Because of this change and specifically [`vm.overcommit_memory`](https://github.com/open-contracting/deploy/blob/main/salt/postgres/init.sls#L82-L84), the memory requirements on this server has increased to needing at least 6GB SWAP / 4GB RAM (from 2GB SWAP / 4GB RAM)

Few tasks are remaining before this is merged in:
- [ ] Update Documentation for onboarding.sls change
- [ ] Import the redash database and test the site
- [ ] Configure networking for the new IPv6 linode block #278
- [ ] Update the redash set up docs https://ocdsdeploy.readthedocs.io/en/latest/deploy/redash.html

@jpmckinney  / @yolile , I can see redash is configured to send email through SES, what is this used for? If I import the database from ocp08 for testing will it trigger any emails?

jpmckinney commented on salt/top.sls:
```diff
@@ -48,6 +48,7 @@ base:
 
   'redash':
     - redash
+    - postgres
```
I figure this should be first?

RobHooper replied:
Doesn't matter for the redash application but swapped order so it is alphabetical.

jpmckinney commented on salt/redash/init.sls:
```diff
@@ -1,16 +1,24 @@
+    - mode: 644
+    - template: jinja
+
+{{ create_pg_database( salt['pillar.get']('redash:postgres:database'), salt['pillar.get']('redash:postgres:username') ) }}
```
It's unusual to use `pillar.get` here. If the configurations are optional, then we should put an if-statement around the call to the `create_pg_database` macro – otherwise, we're running the macro with null values.

RobHooper replied:
Updated to meet pillar.get standards matching docs ✔

jpmckinney commented on salt/redash/init.sls:
```diff
@@ -1,16 +1,24 @@
-
-/opt/redash:
-  file.directory:
+# Create Redash configuration and directories
```
Is it possible/appropriate to user `docker_apps` like in `registry/init.sls` ? If so, then the env file would instead be configured mostly in Pillar directly.

I haven't yet written up the docs page for it. #309

RobHooper replied:
I have updated redash to use `docker_apps`.

jpmckinney commented on salt/redash/init.sls:
```diff
@@ -1,16 +1,24 @@
 include:
   - docker
 
 # =================================================================================================
 # See https://ocdsdeploy.readthedocs.io/en/latest/deploy/redash.html for installation instructions.
 # =================================================================================================
 
```
I think this can now be removed? (And the linked docs updated.)

RobHooper replied:
Header removed.

jpmckinney commented on salt/redash/files/env:
```diff
@@ -0,0 +1,14 @@
+PYTHONUNBUFFERED=0
+REDASH_LOG_LEVEL=INFO
+REDASH_REDIS_URL=redis://redis:6379/0
+REDASH_COOKIE_SECRET={{ salt['pillar.get']('redash:cookie_secret') }}
+REDASH_SECRET_KEY={{ salt['pillar.get']('redash:secret_key') }}
+REDASH_DATABASE_URL=postgresql://{{ salt['pillar.get']('redash:postgres:username') }}:{{ salt['pillar.get']('redash:postgres:password') }}@{{ salt['pillar.get']('redash:postgres:hostname') }}/{{ salt['pillar.get']('redash:postgres:database') }}
+REDASH_FEATURE_SHOW_PERMISSIONS_CONTROL=true
+REDASH_MAIL_SERVER={{ salt['pillar.get']('redash:mail:server') }}
+REDASH_MAIL_PORT={{ salt['pillar.get']('redash:mail:port') }}
+REDASH_MAIL_USE_TLS=true
+REDASH_MAIL_USERNAME={{ salt['pillar.get']('redash:mail:username') }}
+REDASH_MAIL_PASSWORD={{ salt['pillar.get']('redash:mail:password') }}
+REDASH_MAIL_DEFAULT_SENDER={{ salt['pillar.get']('redash:mail:default_sender') }}
+REDASH_HOST={{ salt['pillar.get']('redash:host') }}
```
Use `pillar.key1.key2` instead of `salt['pillar.get']('key1:key2')`, unless the value is truly optional.

See https://ocdsdeploy.readthedocs.io/en/latest/develop/style/language.html#jinja-templating-language

jpmckinney commented on salt/redash/files/env:
```diff
@@ -0,0 +1,14 @@
+PYTHONUNBUFFERED=0
+REDASH_LOG_LEVEL=INFO
+REDASH_REDIS_URL=redis://redis:6379/0
```
Would it be relevant to pull Redis out of Docker, like PostgreSQL? Just wondering. I'm fine to leave it in.

jpmckinney commented on salt/redash/files/env:
```diff
@@ -0,0 +1,14 @@
+REDASH_REDIS_URL=redis://redis:6379/0
+REDASH_COOKIE_SECRET={{ salt['pillar.get']('redash:cookie_secret') }}
+REDASH_SECRET_KEY={{ salt['pillar.get']('redash:secret_key') }}
+REDASH_DATABASE_URL=postgresql://{{ salt['pillar.get']('redash:postgres:username') }}:{{ salt['pillar.get']('redash:postgres:password') }}@{{ salt['pillar.get']('redash:postgres:hostname') }}/{{ salt['pillar.get']('redash:postgres:database') }}
```
The current server also sets `POSTGRES_PASSWORD` but not sure if it's actually used by Redash.

RobHooper replied:
This variable is not used AFAICS.

jpmckinney commented on salt/postgres/files/conf/redash.conf:
```diff
@@ -0,0 +1,5 @@
+
+data_directory = '/var/lib/postgresql/13/main'
+
+shared_buffers = 128MB
```
Why this value?

jpmckinney commented on salt/onboarding.sls:
```diff
@@ -10,16 +10,14 @@ update all packages:
 
-{%- if 'fqdn_ip4' in grains %}
-{{ grains.fqdn_ip4[0] }}:
+{{ pillar.main_ipv4 }}:
```
This means adding IPs for all servers in Pillar?

RobHooper replied:
I have changed how this works, refer to new server onboarding and netplan docs.

jpmckinney commented on salt-config/roster:
```diff
@@ -8,11 +8,13 @@ docs:                ocp07.open-contracting.org
 spoonbill-dev:       ocp12.open-contracting.org
 toucan:              ocp09.open-contracting.org
 
+# ocp00 is the Redmine CRM server
+# ocp08 was redash on Ubuntu 18
```
```suggestion
# ocp0 is crm.open-contracting.org
# ocp08 was redash.open-contracting.org on Ubuntu 18
```

jpmckinney commented on pillar/redash.sls:
```diff
@@ -9,5 +9,26 @@ apache:
+  docker_tag: redash/redash:10.0.0-beta.b49597
+  postgres:
+    username: redash_user
+    hostname: ocp14.open-contracting.org
```
Why can't it connect on localhost?

Some docker-compose files use `network_mode: "host"`, but I know this has its own disadvantages.

jpmckinney reviewed (commented):
For the docker-compose file, the differences are:

* remove postgres
* upgrade redis and add healthcheck
* change docker-compose file version
* change docker image tag

RobHooper commented:
@jpmckinney , this PR is ready to review. 

I have...
* Updated redash to use `docker_apps`
* Added netplan configuration enabling us to configure IPv6 on Linode servers
* With the netplan changes, I have added configs to registry addressing issue #278 
* Updated documentation around onboarding and created docs for netplan
* Updated redash version for the new stable release
* Updated redash documentation for new setup and processes.

I have also imported the database from the old server onto ocp14 and tested, confirming the app is working fine :)

RobHooper commented:
Noting migration steps:

1. Merge PR and changes in private repo
    - https://github.com/open-contracting/deploy-pillar-private/pull/24
2. Follow Documented ["Initialize Redash" process](https://github.com/open-contracting/deploy/blob/c4038b9f1f83616544a1182657410e5432f04f00/docs/deploy/redash.rst#initialize-redash)
3. Copy over database from OCP with one additional step:
```bash
sed -i 's/OWNER TO postgres/OWNER TO redash_user/g; /COMMENT ON DATABASE postgres/d' redash.sql
```
4. Update DNS

RobHooper commented on salt/redash.sls:
```diff
@@ -0,0 +1,6 @@
+include:
+  - docker_apps
+
+{{ create_pg_database( "redash_db", "redash_user" ) }}
```
@jpmckinney I see the `create_pg_database` macro has been removed, how is this now configured?

RobHooper replied:
I've modified this to use the new system in `salt/postgres/init.sls` / putting the databases in pillar.
Might need some tweaking, I couldn't find any documentation or use of this new set up yet.

jpmckinney replied:
It's step 2 here. I'll fix the heading for the section to make it clearer.

jpmckinney replied:
https://ocdsdeploy.readthedocs.io/en/latest/develop/update/postgres.html#add-service-accounts

RobHooper replied:
Happy the PR is updated for new postgres install + configuration process.

jpmckinney commented on salt/core/network/init.sls:
```diff
@@ -0,0 +1,61 @@
+    - order: 5
+    - contents: "{{ pillar.host_id }}.open-contracting.org"
+
+# The Salt system.networking state does not fully support Ubuntu 20.04 yet so we are using cmd.run instead.
```
Can you provide a link to an issue?

jpmckinney replied:
Aha, https://github.com/open-contracting/deploy/issues/278#issuecomment-924485063.

jpmckinney commented on pillar/redash.sls:
```diff
@@ -1,3 +1,19 @@
+    gateway_ip: 139.162.199.1
+    dns_servers: [ 178.79.182.5, 176.58.107.5, 176.58.116.5, 176.58.121.5, 151.236.220.5, 212.71.252.5, 212.71.253.5, 109.74.192.20, 109.74.193.20, 109.74.194.20 ]
+  ipv6:
+    primary_ip: 2a01:7e00:e000:02cc::14
```
Where is this from? I don't see any IP ending in `::14` at https://cloud.linode.com/linodes/30254827/networking

Also, why don't we need to set a `primary_ip_subnet_mask`? It defaults to 128, but I thought our mask was 64:

```
Open a support ticket with Linode to add the new server to our ``2a01:7e00:e000:02cc::/64`` IPv6 range.
```

RobHooper replied:
We have all IP addresses under `2a01:7e00:e000:02cc:...` and this is for our entire Linode network - shared across all severs.
I chose `::14` because this is running on OCP14.

I only expect to only require one IPv6 address (`/128`) per server most of the time.

jpmckinney replied:
Aha, we can update the docs to explain how to assign IP addresses without duplication.

jpmckinney commented on docs/deploy/create_server.rst:
```diff
@@ -16,35 +16,48 @@ Linode
+   #. Copy *SSH Access* to your clipboard
 
-#. Select your new server and copy the *SSH Access* details
+#. Open a support ticket with Linode to add the new server to our ``2a01:7e00:e000:02cc::/64`` IPv6 range.
```
Where/how is this IPv6 range requested/managed?

RobHooper replied:
I created a ticket with Linode asking them to provision a /64 block to us.
https://cloud.linode.com/support/tickets/16265355

We need a /64 block because this is the resolution SPAM blocklists work on. Having the whole range this protects us from "noisy" neighbours sending spam.

At the moment IPv6 management in Linode is entirely done by their support team via support tickets. 

I'll add a link to the Linode support portal to these docs so it is to hand.
https://cloud.linode.com/support/tickets

jpmckinney commented on pillar/registry.sls:
```diff
@@ -1,3 +1,29 @@
+network:
+  host_id: ocp13
```
Right now mailname is amy01.open-contracting.org and host files have both ocp13 and amy01. Do we want to change it with this commit?

RobHooper replied:
Agreed, let's make this change on it's own PR.
This PR sets the groundwork for #278 but doesn't have to implement it.

jpmckinney commented on docs/develop/update/network.rst:
```diff
@@ -0,0 +1,85 @@
+
+``ipv4``
+  The *Address* with a *Type* of *IPv4 – Public*
+``ipv6``
+  TODO
+``ipv4_subnet_mask``
+  TODO
+``ipv6_subnet_mask``
+  TODO
```
@RobHooper Can you fill in the TODOs I added here? I don't know where to find these values from Linode.

RobHooper replied:
I've simplified the config, removing some of the values we're not likely going to change (subnet_mask, nameservers and search).
I've filled in the TODO's for the remaining items.

jpmckinney commented on docs/develop/update/network.rst:
```diff
@@ -0,0 +1,89 @@
+``nameservers.addresses``
+  The *DNS Resolvers*
+``nameservers.search``
+  Default ``[ open-contracting.org ]``
```
Please document in what case the default would be overridden.

RobHooper replied:
I can't see a time when these would be overwritten by us.
I've removed these as configurable pillar items simplifying the config.

jpmckinney commented on pillar/redash.sls:
```diff
@@ -1,3 +1,17 @@
+    configuration: linode
+    ipv4_subnet_mask: "/24"
+    addresses:
+      - 2a01:7e00::f03c:92ff:fea5:0e5f/128  # SLAAC
```
Why is /128 added to the SLAAC IP? I only see the IP (not the mask) on Linode.

RobHooper replied:
I think I assumed it was a single IP address which should be indicated with `/128`.
Netplan requires a subnet_mask to be defined.

Looking again I see the Linode Networking page shows a subnet mask of `ffff:ffff:ffff:ffff::` which is `/64`.
This is also in the Linode documentation: https://www.linode.com/docs/guides/linux-static-ip-configuration/#ubuntu-1804-and-2004

Both have worked in my testing - we are not using SLAAC since we are using static configuration but we have to set it up for netplan.

RobHooper replied:
I'll update this to /64.

jpmckinney commented on pillar/redash.sls:
```diff
@@ -6,7 +20,34 @@ apache:
+  type: oltp
+  nr_hugepages: 128
+  databases:
+    redash_db:
```
FYI, instead of having every database be named `X_db`, `Y_db`, etc., we just name database `X`, `Y`, etc.

jpmckinney commented on docs/deploy/redash.rst:
```diff
@@ -66,110 +47,86 @@ Run script
 
-      .. code-block:: bash
+      cd /data/deploy/redash/
+      docker-compose up -d
```
We're starting the containers as root? Shouldn't we have a non-root user?

jpmckinney replied:
Aha, we do have a `deployer` user, so I'll add `su deployer`.

jpmckinney commented on docs/deploy/create_server.rst:
```diff
@@ -190,11 +203,13 @@ Hostnames follow the format ``ocp##.open-contracting.org`` (ocp01, ocp02, etc.).
    .. code-block:: bash
 
-      salt-ssh --log-level=trace TARGET state.apply 'onboarding,core*' pillar='{"host_id": "ocpXX"}'
+      salt-ssh --log-level=trace TARGET state.apply 'onboarding,core*'
```
Should the create_server process also link to https://ocdsdeploy.readthedocs.io/en/latest/develop/update/firewall.html?

RobHooper replied:
Yes, I've just added a "Firewall" step linking to these docs. 
It should only be needed for docker deployments.

jpmckinney commented on pillar/redash.sls:
```diff
@@ -6,7 +20,34 @@ apache:
+postgres:
+  version: 13
+  # Redash connections to Postgres are not sent over the localhost connection.
+  # Postgres access is locked down further in the Linode firewall.
```
Does this involve more than what's currently documented at https://ocdsdeploy.readthedocs.io/en/latest/develop/update/firewall.html#linode-vps-servers ?

RobHooper replied:
The ocp14 Linode firewall has a allow line granting Postgres access to it's own IPs.
I'm not sure where it would be best to document this? perhaps on the [redash page?](https://ocdsdeploy.readthedocs.io/en/latest/deploy/redash.html#create-a-redash-server)
This is a docker quirk mainly.

RobHooper replied:
Reviewing the firewall we removed the requirement for the complex "allow self" line now we've got host.docker.local working as the database connection. So the firewall is much simpler. 
I've added this to the redash page as well.

jpmckinney commented on pillar/redash.sls:
```diff
@@ -6,7 +20,34 @@ apache:
+
+postgres:
+  version: 13
+  # Redash connections to Postgres are not sent over the localhost connection.
```
Can you remind me why this is?

jpmckinney replied:
If I remember correctly, it's because Docker uses the external IP.

RobHooper replied:
Correct - Docker is connecting to Postgres in a strange way.

Containers are running in their own network on the `172.17.0.0/16` range (at least by default).
The Redash containers are not aware of any network outside of itself, additionally "localhost" is the container not the server.

jpmckinney commented on salt/docker_apps/files/redash.yaml:
```diff
@@ -0,0 +1,57 @@
+# Based on: https://github.com/getnewdash/setup/blob/master/data/docker-compose.yml
+version: "2.1"
+x-redash-service: &redash-service
+  image: redash/redash:10.0.0.b50363
```
Should we use 10.1.0.b50633 https://hub.docker.com/r/redash/redash/tags ?

RobHooper commented on pillar/redash.sls:
```diff
@@ -29,6 +29,8 @@ postgres:
   # Docker containers don't use localhost to connect to the host's PostgreSQL service. Public access is controlled using Linode's firewall.
   public_access: True
   configuration: redash
+  storage: ssd
+  type: oltp
```
True, I'm happy either way. We can remove the explicit definition if you prefer.

jpmckinney commented on salt/core/network/files/netplan_linode.yaml:
```diff
@@ -0,0 +1,23 @@
+      dhcp6: no
+      accept-ra: no
+      addresses:
+	{#- networkd makes the latest IP the primary IP so order matters here #}
```
I assume "latest" means "last"? I changed it to avoid ambiguity.

That said, https://www.linode.com/docs/guides/linux-static-ip-configuration/#ubuntu-1804-and-2004 puts the SLAAC address last. It also includes a private IP. I don't know if "IPv6 – Link Local" in Linode matches that private IP or is something else.

RobHooper replied:
> I assume "latest" means "last"? I changed it
✔✔

In my testing I've had *lots* of issues due to the ordering the IPv6 IPs. The config in this PR is the most reliable set up. 
As I remember, putting SLAAC last will prioritise the other IPv6 address (from the Linode pool) over our IPv6 pool.
The server picks up the Link Local address automatically so no need for manual configuration.

jpmckinney replied:
I'm having trouble finding documentation about the priority order. This thread suggests that the first entry is prioritized, not the last: https://bugs.launchpad.net/netplan/+bug/1766656

The manpage for systemd-network doesn't seem to describe any ordering logic: https://manpages.ubuntu.com/manpages/bionic/man5/systemd.network.5.html

jpmckinney replied:
Anyhow, the logic seems to be undocumented/unspecified, so if this config works, let's go ahead with the merge. I've added comments to the file, in case we later discover documented logic.

RobHooper replied:
Agreed - the documentation around Netplan is poor.
I'm happy that this configuration is working fine on OCP14.
I will merge this PR tomorrow morning (20 Apr 2022). :)

jpmckinney reviewed (commented):
Almost there!

RobHooper commented on docs/deploy/create_server.rst:
```diff
@@ -16,35 +16,50 @@ Linode
+   #. Click *Power On*
+   #. Copy *SSH Access* to your clipboard
+
+#. Open a `support ticket with Linode<https://cloud.linode.com/support/tickets>`__ to add the new server to our ``2a01:7e00:e000:02cc::/64`` IPv6 range.
```
We need to raise a Linode support ticket for all IPv6 configuration at the ISP level because they don't provide access to configure IPv6 in their support interface. I assume they need to configure this by hand in their routers + switches.
The only thing we need Linode to action is assigning the IPv6 IP to the new server (from our IPv6 range)
Support ticket template:
> Hello,
> Please can you assign our IPv6 IP range (2a01:7e00:e000:02cc::/64) to our new server ocpXX.open-contracting.org.
> Thank you,

In within the next 24 hours I expect a reply just confirming that they've actioned this and we can proceed / re-run just the network set up portion of salt. :)

jpmckinney replied:
Ok, so each time we’re (1) requesting from Linode that the entire range be assigned to the new server - and then (2) in netplan choosing the specific IP (without any involvement from Linode)?
